keps/sig-scheduling/3370-native-podgroup-api/README.md

```diff
@@ -0,0 +1,1076 @@
+	// an in-progress PodGroup-level scheduling attempt.
+	ScheduleTimeoutSeconds *int32
+}
+
+// Subset represents a collection of pods with the same role.
+type Subset struct {
+	// Name is used to distinguish pods with different roles in the
+	// same pod group. Optional if there is only one subset in the 
+	// pod group.
+	Name string
+
+	// MinMember specifies the minimum number of pods required for a 
+	// subset to be operational.
+	MinMember int32
+}
+```
+
+In order to associate a pod with its target PodGroup, we need to add a field call `PodGroup` in the podSpec to declare which PodGroup and Subset the pod belongs to.
+
+```go
+// PodSpec is a description of a pod.
+type PodSpec struct {
+	...
+	PodGroup PodGroupRef
+	...
+}
+
+// PodGroupRef contains enough information to let you identify an owning
+// pod group.
+// - pod.spec.podGroup.name = foo
+// - pod.spec.podGroup.subset = (bar/baz)
+type PodGroupRef struct {
+  // Name is the name of PodGroup.
+  Name   string
+  // Subset is the name of PodGroup.
+  Subset string
+}
+```
+
+In this design, a PodGroupStatus is needed to record the latest status. This
+enables users to quickly know what’s going on with this PodGroup; also some
+integrators can rely on it to make pending PodGroup schedulable. For this, we 
+need to add a new PodGroup Controller in controller-manager.
+
+```go
+// PodGroupStatus represents the current state of a pod group.
+type PodGroupStatus struct {
+	// Represents the latest available observations of 
+	// a pod group's current state.
+	Conditions []PodGroupCondition
+
+	// Subsets is a list of SubsetStatus which represents 
+	// the current state of a single subset.
+	Subsets []SubsetStatus
+}
+
+// PodGroupCondition describes the state of a pod group 
+// at a certain point.
+type PodGroupCondition struct {
+	// Type of deployment condition.
+	Type PodGroupConditionType
+
+	// Status of the condition, one of True, False, Unknown.
+	Status v1.ConditionStatus
+
+	// The last time this condition was updated.
+	LastUpdateTime metav1.Time
+
+	// Last time the condition transitioned from one status to another.
+	LastTransitionTime metav1.Time
+	
+	// The reason for the condition's last transition.
+	Reason string
+	
+	// A human readable message indicating details about 
+	// the transition.
+	Message string
+}
+
+type PodGroupConditionType string
+
+const (
+	// PodGroupScheduled represents status of the scheduling 
+	// process for this pod group.
+	PodGroupScheduled PodGroupConditionType = "PodGroupScheduled"
+)
+
+// SubsetStatus represents the current state of a subset.
+type SubsetStatus struct {
+	Name string
+
+	// Total number of non-terminated pods targeted by this subset. (
+	// their labels match the selector)
+	Total int32
+
+	// ScheduledAvailable is the number of pods in the subset that can be scheduled
+	// successfully, but the scheduling fails because minMember cannot be met.
+	ScheduledAvailable int32
```

Comment:
No. Not in kube-scheduler. It will be updated by pod-group-controller in kube-controller by the reconcile mechanism.
It will use `pod.condition.reason = CoschedulingNotMeetMinMember` to determine which class the pod belongs to.

Reply:
oh, so for each pod waiting on permit we will still send a pod update?

Reply:
If the pod group times out, it needs to release the resources they are occupying. For this part of pod group which can be scheduled successfully, but the scheduling fails because minMember cannot be met. We need to update the pod.condition as `CoschedulingNotMeetMinMember ` in the func `handleSchedulingFailure`.

If they can meet the minMember, we don't need to update it.